File: src/main/com/engine/entities/contracts/IBodyEntity.java
package com.engine.entities.contracts;

import com.engine.common.ClassInstanceUtils;
import com.engine.common.interfaces.IPositional;
import com.engine.common.interfaces.ISizable;
import com.engine.entities.IGameEntity;
import com.engine.world.Body;
import com.engine.world.BodyComponent;
import kotlin.reflect.KClass;

import java.util.Objects;

/**
 * Interface for entities that have a body.
 */
public interface IBodyEntity extends IGameEntity, IPositional, ISizable {

    default Body getBody() {
        KClass<BodyComponent> key = ClassInstanceUtils.convertToKClass(BodyComponent.class);
        return Objects.requireNonNull(getComponent(key)).getBody();
    }

    default float getX() {
        return getBody().getX();
    }

    default float getY() {
        return getBody().getY();
    }

    default void setX(float x) {
        getBody().setX(x);
    }

    default void setY(float y) {
        getBody().setY(y);
    }

    default void setPosition(float x, float y) {
        getBody().setPosition(x, y);
    }

    default float getWidth() {
        return getBody().getWidth();
    }

    default float getHeight() {
        return getBody().getHeight();
    }

    default void setWidth(float width) {
        getBody().setWidth(width);
    }

    default void setHeight(float height) {
        getBody().setHeight(height);
    }

    default void setSize(float width, float height) {
        getBody().setSize(width, height);
    }

    default void translateSize(float width, float height) {
        getBody().setWidth(getBody().getWidth() + width);
        getBody().setHeight(getBody().getHeight() + height);
    }

    default void translate(float x, float y) {
        getBody().translation(x, y);
    }
}

File: core/src/main/com/engine/entities/contracts/IBodyEntity.kt
package com.engine.entities.contracts

import com.engine.common.interfaces.Positional
import com.engine.common.interfaces.Sizable
import com.engine.entities.IGameEntity
import com.engine.world.Body
import com.engine.world.BodyComponent

interface IBodyEntity : IGameEntity, Positional, Sizable {

  val body: Body
    get() = getComponent(BodyComponent::class)!!.body

  override fun getX() = body.x

  override fun getY() = body.y

  override fun getPosition() = body.getPosition()

  override fun setPosition(x: Float, y: Float) {
    body.setPosition(x, y)
  }

  override fun getWidth() = body.width

  override fun getHeight() = body.height

  override fun setSize(width: Float, height: Float) {
    body.setSize(width, height)
  }

  override fun setWidth(width: Float) {
    body.setWidth(width)
  }

  override fun setHeight(height: Float) {
    body.setHeight(height)
  }

  override fun translateSize(width: Float, height: Float) {
    body.setWidth(body.width + width)
    body.setHeight(body.height + height)
  }

  override fun translate(x: Float, y: Float) {
    body.translation(x, y)
  }
}
